### PR TITLE
$dynamicRef - search scoped resources for $dynamicAnchor

### LINKED ISSUE #3: Failing $dynamicRef cases on main
Running pytest, 2 failures occur:

```
FAILED tests/test_suite.py::test_validate[2020-12 -> dynamicRef.json -> A $dynamicRef should resolve to the first $dynamicAnchor that is encountered when the schema is evaluated -> An array containing non-strings is invalid]
FAILED tests/test_suite.py::test_validate[2020-12 -> dynamicRef.json -> A $dynamicRef with intermediate scopes that don't include a matching $dynamicAnchor should not affect dynamic scope resolution -> An array containing non-strings is invalid]
```

### PR DESCRIPTION
**Pull request by rjmill**
Instead of looking for a "$dynamicAnchor" keyword in each schema in the
dynamic scope, look at anchors defined in each scoped schema *resource*.

The 2020-12 spec says `$dynamicRef` (when acting dynamically) replaces the
initially resolved URI with "the outermost schema resource in the
dynamic scope that defines an identically named fragment with
'$dynamicAnchor'."

The wording feels ambiguous, but the official test suite clarifies the
expectation:
Unlike the `$recursive*` keywords (which were looking for the "outermost
schema" that *contains* a specific keyword), `$dynamic*` are looking for
the outermost schema *resource* that *defines* a specific anchor.

tl;dr Official test suite cases were failing. The wording of the spec is
not great, but (as evidenced by this commit message) I can't find a
clearer way to word it.

Fixes #3

### CODE REVIEW

**Comment by marksparkza** [comment]
Thanks for the PR @rjmill. Your logic is essentially correct - it fixes the failing cases and works for the other $dynamicRef tests.

There were, however, a number of issues with my $dynamicRef implementation, so I have basically rewritten it - see 9775dabe20f4600d54e2786d19cf6443f8b17e46 for the changes.

Thanks again for your input on this!
